File: packages/ts/experiments/src/runner.ts
import cliProgress from "cli-progress";
import { enableExperimentMode, registerOTel } from "@uselemma/tracing";
import { LemmaExperimentsClient } from "./client";
import type {
  ExperimentResult,
  ExperimentSummary,
  LemmaExperimentRunnerOptions,
  RunExperimentOptions,
  TestCase,
} from "./types";

export class LemmaExperimentRunner {
  private readonly client: LemmaExperimentsClient;
  private readonly tracerProvider: { forceFlush(): Promise<void> };

  constructor(options: LemmaExperimentRunnerOptions = {}) {
    this.tracerProvider = registerOTel({
      apiKey: options.apiKey ?? process.env.LEMMA_API_KEY,
      projectId: options.projectId ?? process.env.LEMMA_PROJECT_ID,
      baseUrl: options.baseUrl,
    });
    enableExperimentMode();

    this.client = new LemmaExperimentsClient({
      apiKey: options.apiKey ?? process.env.LEMMA_API_KEY,
      baseUrl: options.baseUrl ?? process.env.LEMMA_API_URL,
    });
  }

  async runExperiment<TInput = Record<string, unknown>>(
    options: RunExperimentOptions<TInput>
  ): Promise<ExperimentSummary> {
    const {
      experimentId,
      strategyName,
      agent,
      concurrency,
      progress = true,
    } = options;

    const testCases = await this.client.getTestCases(experimentId);
    const total = testCases.length;

    const results: ExperimentResult[] = [];
    let completed = 0;

    const maybeIncrement = () => {
      completed++;
      if (progress && bar) {
        bar.update(completed);
      }
    };

    let bar: cliProgress.SingleBar | null = null;
    if (progress && total > 0) {
      bar = new cliProgress.SingleBar(
        {},
        cliProgress.Presets.shades_classic
      );
      bar.start(total, 0);
    }

    const runOne = async (testCase: TestCase<TInput>): Promise<ExperimentResult | null> => {
      try {
        const { runId } = await agent(testCase.inputData as TInput);
        const result: ExperimentResult = { runId, testCaseId: testCase.id };
        results.push(result);
        return result;
      } catch {
        return null;
      } finally {
        maybeIncrement();
      }
    };

    const cases = testCases as TestCase<TInput>[];
    if (concurrency != null && concurrency > 0) {
      await this.runWithConcurrency(cases, runOne, concurrency);
    } else {
      await Promise.all(cases.map(runOne));
    }

    if (bar) {
      bar.stop();
    }

    await this.tracerProvider.forceFlush();
    await this.client.recordResults(experimentId, strategyName, results);

    return { successful: results.length, total };
  }

  private async runWithConcurrency<T>(
    items: T[],
    fn: (item: T) => Promise<unknown>,
    concurrency: number
  ): Promise<void> {
    let index = 0;
    async function worker(): Promise<void> {
      while (true) {
        const i = index++;
        if (i >= items.length) break;
        await fn(items[i]);
      }
    }
    await Promise.all(
      Array.from(
        { length: Math.min(concurrency, items.length) },
        () => worker()
      )
    );
  }
}

File: packages/py/experiments/src/uselemma_experiments/runner.py
"""LemmaExperimentRunner - orchestrate experiment runs."""

from __future__ import annotations

import asyncio
import os
from collections.abc import Awaitable, Callable
from typing import Any

from uselemma_tracing import enable_experiment_mode, register_otel

from .client import LemmaExperimentsClient
from .types import ExperimentResult, ExperimentSummary, TestCase


class LemmaExperimentRunner:
    """Run experiments: fetch test cases, run agent, flush traces, record results."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        project_id: str | None = None,
        base_url: str | None = None,
    ) -> None:
        api_key = api_key or os.environ.get("LEMMA_API_KEY")
        project_id = project_id or os.environ.get("LEMMA_PROJECT_ID")
        base_url = base_url or os.environ.get("LEMMA_API_URL") or "https://api.uselemma.ai"

        self._tracer_provider = register_otel(
            api_key=api_key,
            project_id=project_id,
            base_url=base_url,
        )
        enable_experiment_mode()

        self._client = LemmaExperimentsClient(
            api_key=api_key,
            base_url=base_url,
        )

    async def run_experiment(
        self,
        experiment_id: str,
        strategy_name: str,
        agent: Callable[[dict[str, Any]], Awaitable[dict[str, Any]]],
        *,
        concurrency: int | None = None,
        progress: bool = True,
    ) -> ExperimentSummary:
        """Run the agent on all test cases, flush traces, and record results."""
        test_cases = await self._client.get_test_cases(experiment_id)
        total = len(test_cases)
        results: list[ExperimentResult] = []

        async def run_one(test_case: TestCase) -> ExperimentResult | None:
            try:
                out = await agent(test_case["inputData"])
                run_id = out.get("runId") or out.get("run_id")
                if not run_id:
                    return None
                return {"runId": run_id, "testCaseId": test_case["id"]}
            except Exception:
                return None

        if concurrency is not None and concurrency > 0:
            sem = asyncio.Semaphore(concurrency)

            async def bounded_run(test_case: TestCase) -> ExperimentResult | None:
                async with sem:
                    return await run_one(test_case)

            tasks = [bounded_run(tc) for tc in test_cases]
            if progress and total > 0:
                from tqdm.asyncio import tqdm_asyncio

                completed = await tqdm_asyncio.gather(*tasks, desc="Experiment")
            else:
                completed = await asyncio.gather(*tasks)
        else:
            if progress and total > 0:
                from tqdm.asyncio import tqdm_asyncio

                completed = await tqdm_asyncio.gather(
                    *[run_one(tc) for tc in test_cases],
                    desc="Experiment",
                )
            else:
                completed = await asyncio.gather(*[run_one(tc) for tc in test_cases])

        for r in completed:
            if r is not None:
                results.append(r)

        await self._tracer_provider.force_flush()
        await self._client.record_results(experiment_id, strategy_name, results)

        return {"successful": len(results), "total": total}
